File: Python/OneLang/One/Transforms/InferTypesPlugins/ArrayAndMapLiteralTypeInfer.py
from OneLangStdLib import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Interfaces as ints

class ArrayAndMapLiteralTypeInfer(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("ArrayAndMapLiteralTypeInfer")
    
    def infer_array_or_map_item_type(self, items, expected_type, is_map):
        item_types = []
        for item in items:
            if not ArrayHelper.some(lambda t: astTypes.TypeHelper.equals(t, item.get_type()), item_types):
                item_types.append(item.get_type())
        
        literal_type = self.main.current_file.literal_types.map if is_map else self.main.current_file.literal_types.array
        
        item_type = None
        if len(item_types) == 0:
            if expected_type == None:
                self.error_man.warn(f'''Could not determine the type of an empty {("MapLiteral" if is_map else "ArrayLiteral")}, using AnyType instead''')
                item_type = astTypes.AnyType.instance
            elif isinstance(expected_type, astTypes.ClassType) and expected_type.decl == literal_type.decl:
                item_type = expected_type.type_arguments[0]
            else:
                item_type = astTypes.AnyType.instance
        elif len(item_types) == 1:
            item_type = item_types[0]
        elif not (isinstance(expected_type, astTypes.AnyType)):
            self.error_man.warn(f'''Could not determine the type of {("a MapLiteral" if is_map else "an ArrayLiteral")}! Multiple types were found: {", ".join(list(map(lambda x: x.repr(), item_types)))}, using AnyType instead''')
            item_type = astTypes.AnyType.instance
        return item_type
    
    def can_detect_type(self, expr):
        return isinstance(expr, exprs.ArrayLiteral) or isinstance(expr, exprs.MapLiteral)
    
    def detect_type(self, expr):
        # make this work: `<{ [name: string]: SomeObject }> {}`
        if isinstance(expr.parent_node, exprs.CastExpression):
            expr.set_expected_type(expr.parent_node.new_type)
        elif isinstance(expr.parent_node, exprs.BinaryExpression) and expr.parent_node.operator == "=" and expr.parent_node.right == expr:
            expr.set_expected_type(expr.parent_node.left.actual_type)
        elif isinstance(expr.parent_node, exprs.ConditionalExpression) and (expr.parent_node.when_true == expr or expr.parent_node.when_false == expr):
            expr.set_expected_type(expr.parent_node.when_false.actual_type if expr.parent_node.when_true == expr else expr.parent_node.when_true.actual_type)
        
        if isinstance(expr, exprs.ArrayLiteral):
            item_type = self.infer_array_or_map_item_type(expr.items, expr.expected_type, False)
            expr.set_actual_type(astTypes.ClassType(self.main.current_file.literal_types.array.decl, [item_type]))
        elif isinstance(expr, exprs.MapLiteral):
            item_type = self.infer_array_or_map_item_type(list(map(lambda x: x.value, expr.items)), expr.expected_type, True)
            expr.set_actual_type(astTypes.ClassType(self.main.current_file.literal_types.map.decl, [item_type]))
        
        return True

File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/ArrayAndMapLiteralTypeInfer.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\ArrayAndMapLiteralTypeInfer;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\ArrayLiteral;
use OneLang\One\Ast\Expressions\MapLiteral;
use OneLang\One\Ast\Expressions\CastExpression;
use OneLang\One\Ast\Expressions\BinaryExpression;
use OneLang\One\Ast\Expressions\ConditionalExpression;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\AnyType;
use OneLang\One\Ast\AstTypes\TypeHelper;
use OneLang\One\Ast\Interfaces\IType;

class ArrayAndMapLiteralTypeInfer extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ArrayAndMapLiteralTypeInfer");
        
    }
    
    protected function inferArrayOrMapItemType($items, $expectedType, $isMap) {
        $itemTypes = array();
        foreach ($items as $item) {
            if (!\OneLang\Core\ArrayHelper::some($itemTypes, function ($t) use ($item) { return TypeHelper::equals($t, $item->getType()); }))
                $itemTypes[] = $item->getType();
        }
        
        $literalType = $isMap ? $this->main->currentFile->literalTypes->map : $this->main->currentFile->literalTypes->array;
        
        $itemType = null;
        if (count($itemTypes) === 0) {
            if ($expectedType === null) {
                $this->errorMan->warn("Could not determine the type of an empty " . ($isMap ? "MapLiteral" : "ArrayLiteral") . ", using AnyType instead");
                $itemType = AnyType::$instance;
            }
            else if ($expectedType instanceof ClassType && $expectedType->decl === $literalType->decl)
                $itemType = $expectedType->typeArguments[0];
            else
                $itemType = AnyType::$instance;
        }
        else if (count($itemTypes) === 1)
            $itemType = $itemTypes[0];
        else if (!($expectedType instanceof AnyType)) {
            $this->errorMan->warn("Could not determine the type of " . ($isMap ? "a MapLiteral" : "an ArrayLiteral") . "! Multiple types were found: " . implode(", ", array_map(function ($x) { return $x->repr(); }, $itemTypes)) . ", using AnyType instead");
            $itemType = AnyType::$instance;
        }
        return $itemType;
    }
    
    function canDetectType($expr) {
        return $expr instanceof ArrayLiteral || $expr instanceof MapLiteral;
    }
    
    function detectType($expr) {
        if ($expr->expectedType === null) {
            // make this work: `<{ [name: string]: SomeObject }> {}`
            if ($expr->parentNode instanceof CastExpression)
                $expr->setExpectedType($expr->parentNode->newType);
            else if ($expr->parentNode instanceof BinaryExpression && $expr->parentNode->operator === "=" && $expr->parentNode->right === $expr)
                $expr->setExpectedType($expr->parentNode->left->actualType);
            else if ($expr->parentNode instanceof ConditionalExpression && ($expr->parentNode->whenTrue === $expr || $expr->parentNode->whenFalse === $expr))
                $expr->setExpectedType($expr->parentNode->whenTrue === $expr ? $expr->parentNode->whenFalse->actualType : $expr->parentNode->whenTrue->actualType);
        }
        
        if ($expr instanceof ArrayLiteral) {
            $itemType = $this->inferArrayOrMapItemType($expr->items, $expr->expectedType, false);
            $expr->setActualType(new ClassType($this->main->currentFile->literalTypes->array->decl, array($itemType)));
        }
        else if ($expr instanceof MapLiteral) {
            $itemType = $this->inferArrayOrMapItemType(array_map(function ($x) { return $x->value; }, $expr->items), $expr->expectedType, true);
            $expr->setActualType(new ClassType($this->main->currentFile->literalTypes->map->decl, array($itemType)));
        }
        
        return true;
    }
}
